### PR TITLE
test/threadstest.c: Add a test to load providers concurrently

### PR DESCRIPTION
**Pull request by levitte**
If we don't synchronize properly in the core provider code, and build
with a thread sanitizer, this should cause a crash.

### CODE REVIEW

**Comment by levitte** [comment]
This test is expected to fail, and should be fixed by #14354

**Comment by levitte** [comment]
This is now rebased on top of #14372, to (hopefully) demonstrate the fix

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Approving this subject to the tests passing after a rebase when #14354 is merged.

File: test/threadstest.c [hunk 1]
```diff
@@ -473,6 +473,34 @@ static int test_multi(int idx)
     return testresult;
 }
 
+/*
+ * This test attempts to load several providers at the same time, and if
+ * run with a thread sanitizer, should crash if the core provider code
+ * doesn't synchronize well enough.
+ */
+#define MULTI_LOAD_THREADS 3
```
Comment (paulidale) [inline test/threadstest.c:481]: How reliably does this reproduce problems?
Would more threads be more likely to produce the problem or not?
This looks like a _cheap_ test.

For reference: two threads seemed like enough when I was testing.

Comment (levitte) [reply test/threadstest.c:481]: Two threads seemed fine in my manual tests as well, but I increased that a bit for good measure.  Other than that, I frankly have no idea, but am thinking that if we need to update this, we can do that *later*, i.e. doesn't have to be done in this PR.

**Comment by levitte** [comment]
WIth #14354 merged, this PR can be closed

**Pull request closed without merging**